instrumentation/opentelemetry-instrumentation-remoulade/src/opentelemetry/instrumentation/remoulade/__init__.py

```diff
@@ -0,0 +1,135 @@
+
+        activation = trace.use_span(span, end_on_exit=True)
+        activation.__enter__()
+
+        utils.attach_span(
+            self._span_registry, message.message_id, (span, activation)
+        )
+
+    def after_process_message(
+        self, _broker, message, *, result=None, exception=None
+    ):
+        span, activation = utils.retrieve_span(
+            self._span_registry, message.message_id
+        )
+
+        if span is None:
+            # no existing span found for message_id
+            return
+
+        if span.is_recording():
+            span.set_attribute(_MESSAGE_TAG_KEY, _MESSAGE_RUN)
+            span.set_attribute(_MESSAGE_NAME_KEY, message.actor_name)
+            pass
+
+        activation.__exit__(None, None, None)
+        utils.detach_span(self._span_registry, message.message_id)
+
+    def before_enqueue(self, _broker, message, delay):
+        retry_count = message.options.get("retries")
+
+        operation_name = (
+            "remoulade/send"
+            if retry_count is None
+            else f"remoulade/send(retry-{retry_count})"
+        )
+
+        span = self._tracer.start_span(
+            operation_name, kind=trace.SpanKind.PRODUCER
+        )
+
+        if retry_count is not None:
+            span.set_attribute("retry_count", retry_count)
+
+        if span.is_recording():
+            span.set_attribute(_MESSAGE_TAG_KEY, _MESSAGE_SEND)
+            span.set_attribute(
+                SpanAttributes.MESSAGING_MESSAGE_ID, message.message_id
+            )
+            span.set_attribute(_MESSAGE_NAME_KEY, message.actor_name)
+            pass
+
+        activation = trace.use_span(span, end_on_exit=True)
+        activation.__enter__()
+
+        utils.attach_span(
+            self._span_registry,
+            message.message_id,
+            (span, activation),
+            is_publish=True,
+        )
+
+        if "trace_ctx" not in message.options:
+            message.options["trace_ctx"] = {}
+        inject(message.options["trace_ctx"])
+
+    def after_enqueue(self, _broker, message, delay, exception=None):
+        _, activation = utils.retrieve_span(
+            self._span_registry, message.message_id, is_publish=True
+        )
+
+        if activation is None:
+            # no existing span found for message_id
+            return
+
+        activation.__exit__(None, None, None)
+        utils.detach_span(
+            self._span_registry, message.message_id, is_publish=True
+        )
+
+
+class RemouladeInstrumentor(BaseInstrumentor):
+    def instrumentation_dependencies(self) -> Collection[str]:
+        return _instruments
+
+    def _instrument(self, **kwargs):
+        tracer_provider = kwargs.get("tracer_provider")
+
+        self._tracer = trace.get_tracer(__name__, __version__, tracer_provider)
+
+        instrumentation_middleware = InstrumentationMiddleware(self._tracer)
+        broker.get_broker().add_middleware(instrumentation_middleware)
```

Comment:
It seems like you are setting the existing broker. Does this work with the [auto instrumentation](https://opentelemetry.io/docs/instrumentation/python/automatic/)?

Reply:
You are right. I have made changes so that I can pass an extra default middleware that will then be set to any current or new broker. I have tested it locally and it works.